internal/pkg/scorecard/interfaces.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scorecard
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	scplugins "github.com/operator-framework/operator-sdk/internal/pkg/scorecard/plugins"
+	"github.com/spf13/viper"
+
+	scapiv1alpha1 "github.com/operator-framework/operator-sdk/pkg/apis/scorecard/v1alpha1"
+)
+
+type Plugin interface {
+	Name() string
+	Run() scapiv1alpha1.ScorecardOutput
+}
+
+type genericPlugin struct {
+	binPath string
+}
+
+type internalPlugin struct {
+	name       string
+	pluginType scplugins.PluginType
+}
+
+func (p genericPlugin) Name() string {
+	return filepath.Base(p.binPath)
+}
+
+func (p genericPlugin) Run() scapiv1alpha1.ScorecardOutput {
+	// This should never error since we chdir in getPlugins()
+	if err := os.Chdir(viper.GetString(PluginDirOpt)); err != nil {
+		name := fmt.Sprintf("Failed Plugin: %s", filepath.Base(p.binPath))
```

Comment:
Abstract failed/invalid plugin code into functions as its more or less copied in several places.

Reply:
I don't think this can really be abstracted more? There are slight differences between each of the 5 error cases to make the errors clearer to the user.